Updates to adhoc instance creation

wjordan opened the pull request:
Several updates to fix and improve adhoc-instance creation.

I've split each individual update into separate commits so they can be reviewed/blamed/potentially-reverted separately:

- 74b6f25: Refactor adhoc-instance creation to reuse the existing standard infrastructure stack template and bootstrap script, rather than its own parallel 'standalone' template+script. This simplifies the overall surface area for this feature (`bootstrap_standalone.sh.erb`, `cloud_formation_adhoc_standalone.yml.erb`, at the cost of slightly added complexity to the standard stack (several extra conditionals based on `adhoc` environment).
  - This commit also removes the 'Branch off existing AMI for adhoc' function (`ami.yml.erb`). This was less reliable than building from scratch (it was currently failing adhoc builds because of in-place Bundler upgrades as a result of #18132), and presented its own performance issues (related to the [EBS-snapshot initialization performance penalty](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-initialize.html)) making the performance gains not as great as expected anyway.
- 554b887: Increase default instance size from `t2.large` to `t2.xlarge`, since building apps while the server is running now consumes more total RAM than the smaller instance type provides.
- 97e21a4: Refactored `CERTIFICATE_ARN` constant to a method, so this AWS API call is only made on-demand, not for every single stack creation/update/validation.
- e877cd0: Add an extra chef-debug log file `/var/log/chef-bootstrap-debug` containing the Chef [`:doc` formatter](https://github.com/chef/chef/blob/master/lib/chef/formatters/doc.rb) output. This formatter is a little more verbose than the default (base) output, and streams the standard-output of `execute` blocks, so tailing this debug log can be useful to monitor ongoing progress of long-lived execute blocks like `build:cdo`.
- d7b9f83: Use the S3-packaging cache for adhoc apps builds by removing `build_apps: true` and `use_my_apps: true`. Skips building apps from scratch if the contents haven't changed and the package is already built on staging/test, improving start-up time in this case. (The tradeoff is that with the current logic, apps-tests will also be run when contents have changed, which takes longer to build; I might look to improve this in a future PR.)
- 96879b6: Set `"image_optim": false` by default on standalone stacks, to disable image-optimization by default on these instances (because it takes a long time, and isn't needed in most cases).

wjordan commented on aws/cloudformation/bootstrap_chef_stack.sh.erb:
```diff
@@ -24,11 +24,61 @@ INSTANCE_ID=$(curl -s http://169.254.169.254/latest/meta-data/instance-id)
+      "content": "$KEY"
+    }
+  },
+<% end -%>
```
These above additions make the bootstrap script more or less equivalent to the `standalone` version for adhoc instance creation- the new additions are contained within `if environment == 'adhoc'` or `unless @load_balancer` conditionals so there's no change to non-adhoc environments.

sureshc replied:
The adhoc instance I provisioned using these changes presented a self-signed certificate with `cdn-code.org` as the common name.

wjordan commented on aws/cloudformation/cloud_formation_stack.yml.erb:
```diff
@@ -1,26 +1,40 @@
+  unless `git remote show #{origin} 2>&1 | grep \"(up to date)\" | grep \"#{branch}\" | wc -l`.strip.to_i > 0
+    raise 'Current adhoc branch needs to be up-to-date with GitHub branch of the same name, otherwise deploy will fail.
+To specify an alternate branch name, run `rake adhoc:start branch=BRANCH`.'
+  end
```
This was brought in from the standalone template. I added the `rack_env?(:adhoc)` conditional so non-adhoc builds aren't affected.

wjordan commented on aws/cloudformation/cloud_formation_stack.yml.erb:
```diff
@@ -399,9 +419,14 @@ Resources:
+<%   else -%>
+          TTL: 60
+          ResourceRecords: [!GetAtt Daemon.PublicIp]
+<%   end -%>
```
For standalone adhoc instances without a load-balancer, the DNS entries point directly to the standalone `daemon` EC2 instance, so the added conditions are needed for this case.

wjordan commented on lib/cdo/aws/cloud_formation.rb:
```diff
@@ -15,27 +15,19 @@
-      list_certificates(certificate_statuses: ['ISSUED']).
-      certificate_summary_list.
-      find {|cert| cert.domain_name == "*.#{DOMAIN}" || cert.domain_name == DOMAIN}.
-      certificate_arn
```
Refactored this constant to a method so this AWS API call is only made on-demand, not for every single stack creation/update/validation.

wjordan commented:
added a commit to fix the SSL certificate in the new adhoc stack.

sureshc commented:
Do I understand correctly that an adhoc environment (without CDN or Full Stack enabled) will use the one EC2 instance as both the daemon and the web server?  If so, can the Logical ID of this Resource be dynamically named 'WebServer' if environment is adhoc?

https://github.com/code-dot-org/code-dot-org/blob/7e1d328c95927bcc558d4f41369036dcdd50ffa1/aws/cloudformation/cloud_formation_stack.yml.erb#L488

wjordan commented:
> Do I understand correctly that an adhoc environment (without CDN or Full Stack enabled) will use the one EC2 instance as both the daemon and the web server? 

Yes, a single EC2 instance is used for both CI-build tasks and for serving the HTTP web-application in the basic-adhoc case.

> If so, can the Logical ID of this Resource be dynamically named 'WebServer' if environment is adhoc?

This is possible, I've added a commit for this. The 5 places where `Daemon` was defined/referenced in the stack template are now replaced with `<%= daemon %>` which will dynamically use the alternate/legacy `WebServer` resource name based on the standalone-adhoc condition.

wjordan commented:
Tested/verified that the basic adhoc instance creation (`rake adhoc:start`) is working with the current code. I still want to do more testing/verification on modified apps builds with the s3-package changes, but I'm going to merge this now as-is to unblock basic adhoc-instance builds and apply any further changes to a followup PR.
